types/unist/index.d.ts

```diff
@@ -34,13 +33,23 @@ export interface Node<TData extends object = Data> {
 }
 
 /**
- * Information associated by the ecosystem with the node.
- * Space is guaranteed to never be specified by unist or specifications
+ * Data represents information associated by the ecosystem with the node.
+ *
+ * This space is guaranteed to never be specified by unist or specifications
  * implementing unist.
+ *
+ * This type can be augmented to register custom `data` types.
+ *
+ * @example
+ * declare module 'unist' {
+ *   interface Data {
+ *     // `someNode.data.myId` is typed as `number | undefined`
+ *     myId?: number | undefined
+ *   }
+ * }
  */
-export interface Data {
-    [key: string]: unknown;
-}
+// tslint:disable-next-line
+export interface Data {}
```

Comment:
I specifically removed the record fallback. As we’re in TS, the goal is for everyone to type everything.

My goal is for an `hast.Data` to exist which includes all of `unist.Data`, same for hast/xast/nlcst

Reply:
I have my doubts about this one. Is there ever going to be a data property that is applicable to all unist nodes?

Reply:
I could imagine an `id`, sort of like a `key` from VDOMs.

One thing I’ve been vaguely wondering about is how nodes could come from different files, nodes could have some info leading back to which file they came from?

Regardless, we already have data, so what do you propose? Removing it?
We have interfaces everywhere. The point is to remove the generics and implement this in mdast/hast/xast/nlcst and utilities!

Reply:
I was fine with the solution we had using generics, but also with just specifying it as `Record<string, unknown>` and removing the generic and this type.

Nodes that extend `Node` can just override the `data` property. E.g.:

```ts
interface MdxEsm extends Node {
  type: 'mdxEsm',

  data: {
    program: Program
  }
}
```

I’m leaning towards just removing it.

Reply:
Please see [my commit message for more info](https://github.com/DefinitelyTyped/DefinitelyTyped/pull/65789/commits/e8f0962465519d2ef410da2d6ffa2156135b2ef7). Your proposal does not work with utilities that add to nodes, such as the use cases for `mdast-util-to-hast`.

Reply:
@remcohaszing Ping! :)

Reply:
If we have a use case for adding data to unist nodes globally, I’m all for having the interface the way you have defined it right now. :)

Note that it’s also possible to combine generics with this solution.

```ts
// unist

interface Node<T = object> {
  type: string
  data: Data & T
}

interface Data {}

// mdast-util-to-hast

interface Data {
  hName: string
}

// mdx

interface MdxEsm extends Node<{ estree: {type: 'Program'} }> {
  type: 'mdxEsm'
}
```

as opposed to:

```ts
// unist

interface Node {
  type: string
  data: Data
}

interface Data {}

// mdast-util-to-hast

interface Data {
  hName: string
}

// mdx

interface MdxEsmData extends Data {
  estree: {type: 'Program'}
}

interface MdxEsm extends Node {
  type: 'mdxEsm'
  data: MdxEsmData
}
```

Although the latter isn’t too bad either.

Reply:
I indeed like the latter. It supports the needs of `mdast-util-to-hast`, which as far as I understand generics do not.

I want to make a chain. EsmData <- MdastData <- UnistData.

`UnistData` is probably not particularly useful, but it is specified: https://github.com/syntax-tree/unist#data.
Two example use cases are in  https://github.com/DefinitelyTyped/DefinitelyTyped/pull/65789#discussion_r1231098799.

But I think it’s the same as `Node` from unist: that is never used either.
It’s only used in the implementation chain: EsmNode <- MdastNode <- UnistNode.
We don’t type `Node` as `Record<string, unknown>` either?